language/compiler/bytecode_source_map/src/mapping.rs

```diff
@@ -0,0 +1,486 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::source_map::{ModuleSourceMap, SourceName};
+use failure::prelude::*;
+use libra_types::identifier::IdentStr;
+use std::collections::{BTreeMap, HashMap};
+use vm::access::ModuleAccess;
+use vm::file_format::{
+    CodeOffset, CompiledModule, CompiledScript, FieldDefinitionIndex, FunctionDefinition,
+    FunctionDefinitionIndex, SignatureToken, StructDefinition, StructDefinitionIndex,
+    StructFieldInformation, TableIndex, TypeSignature,
+};
+
+/// An object that associates source code with compiled bytecode and source map.
+#[derive(Debug)]
+pub struct SourceMapping {
+    // The various options that we can set for disassembly.
+    options: SourceMappingOptions,
+
+    // The source code for the bytecode. This is not required for disassembly, but it is required
+    // for being able to print out corresponding source code for marked functions and structs.
+    // Unused for now, this will be used when we start printing function/struct markings
+    _source_code: Option<String>,
+
+    // The source map for the bytecode made w.r.t. to the `source_code`
+    source_map: ModuleSourceMap,
+
+    // The resulting bytecode from compiling the source map
+    bytecode: CompiledModule,
+
+    // Any function markings
+    function_marks: HashMap<TableIndex, FunctionMarking>,
```

Comment:
It might might sense to have an immutable `SourceMapping` that associates bytecode with the source + source map and having a separate `MarkedSourceMapping` type that wraps an immutable source mapping + can be mutated (but only to touch the markings).

Reply:
I like this idea. There's still a lot of work left in marking and printing marked functions. How about I add this in a separate diff?